### ISSUE TITLE
@RestController returns a result when the interface is {}

### ISSUE THREAD

**Issue by iwangjie** [spring-projects/spring-framework#24234]
**Affects:** \ 4.2.5

I defined a method in Controller that returns a Page interface, but in fact I return a PageImpl implementation class object, but I receive a response value of {}

Postman Request result：
![image](https://user-images.githubusercontent.com/23075587/71227072-0435a300-2319-11ea-8d3c-3354a87f370d.png)

Controller RequestMapping:
![image](https://user-images.githubusercontent.com/23075587/71227127-2c250680-2319-11ea-9489-c0072b881a0b.png)

Page:
```java
public interface Page<T> {

    int getTotalPages();

    int getNumber();

    int getSize();

    int getNumberOfElements();

    List<T> getContent();

    boolean hasContent();

    boolean isFirst();

    boolean isLast();

    boolean hasNext();

    boolean hasPrevious();

    long getTotalElements();

}
```

PageImpl:

```java 
public class PageImpl<T> implements Page<T>,Serializable {
    private static final long serialVersionUID = 2693326466571693579L;
    private List<T> content = new ArrayList();
    private long total;
    private PageRequest pageable;

    public PageImpl(){

    }

    public PageImpl(List<T> content, PageRequest pageable, long total){
        this.content.addAll(content);
        this.pageable = pageable;
        this.total = !content.isEmpty() && pageable != null && (long)(pageable.getOffset() + pageable.getPageSize()) > total?(long)(pageable.getOffset() + content.size()):total;
    }


    @Override
    public int getTotalPages() {
        return this.getSize() == 0?1:(int)Math.ceil((double)this.total / (double)this.getSize());
    }

    @Override
    public int getNumber() {
        return this.pageable == null?0:this.pageable.getPageNumber();

    }

    @Override
    public int getSize() {
        return this.pageable == null?0:this.pageable.getPageSize();
    }

    @Override
    public int getNumberOfElements() {
        return this.content.size();
    }

    @Override
    public List getContent() {
        return Collections.unmodifiableList(this.content);    }

    @Override
    public boolean hasContent() {
        return !this.content.isEmpty();
    }

    @Override
    public boolean isFirst() {
        return !this.hasPrevious();
    }

    @Override
    public boolean isLast() {
        return !this.hasNext();
    }

    @Override
    public boolean hasNext() {
        return this.getNumber() + 1 < this.getTotalPages();
    }

    @Override
    public boolean hasPrevious() {
        return this.getNumber() > 0;
    }

    @Override
    public long getTotalElements() {
        return this.total;
    }
}
```

I tried analyzing the code and found that in the AbstractMessageConverterMethodProcessor class

![image](https://user-images.githubusercontent.com/23075587/71227318-be2d0f00-2319-11ea-9932-c996519376e7.png)

The Type obtained here is the Type of the interface, not the Type of the implementation class. I analyzed the source code of Gson and found that Gson is a set of attributes obtained based on the byte code of the passed Type. The interface itself does not have any fields. The cause of this problem.

**Comment by iwangjie** [comment]
Attach the help link I submitted at Gson：[https://github.com/google/gson/issues/1635](https://github.com/google/gson/issues/1635)

**Comment by rstoyanchev** [comment]
First, it looks like this was addressed in 5.1 with #21416. This is how `AbstractMessageConverterMethodProcessor` now [resolves the target type](https://github.com/spring-projects/spring-framework/blob/261956fd08fec9f35abbd2a1667e4acbbacdfc31/spring-webmvc/src/main/java/org/springframework/web/servlet/mvc/method/annotation/AbstractMessageConverterMethodProcessor.java#L201).

Second, it seems that even so there is an issue in Gson too. Using the above example with the latest Spring Framework I can see in [GsonHttpMessageConverter#write](https://github.com/spring-projects/spring-framework/blob/261956fd08fec9f35abbd2a1667e4acbbacdfc31/spring-web/src/main/java/org/springframework/http/converter/json/GsonHttpMessageConverter.java#L102) the `Type` passed in is `ParameterizedTypeImpl` and its value is `Page<NewsInformation>`, but it still doesn't render correctly. The `type` comes from the base class `AbstractJsonHttpMessageConverter` which is also used for Jackson JSON and that works as expected for this scenario.

Note that 4.2.x is no longer a supported branch and it is a security risk for you. Upgrading to 4.3.x at least is highly recommended. Obviously you still won't get this feature until you upgrade to 5.1, and then there is the Gson issue still. As a workaround, try using a sub-class the generic type. For example:

```java
public class NewsInformationPageImpl extends Page<NewsInformation> {
}
```
